tensorflow/lite/kernels/reduce.cc

```diff
@@ -740,6 +789,79 @@ TfLiteStatus EvalSum(TfLiteContext* context, TfLiteNode* node) {
 
   return kTfLiteOk;
 }
+
+template <typename T>
+TfLiteStatus EvalQuantizedProd(TfLiteContext* context, TfLiteNode* node,
+                               OpContext* op_context) {
+  OpData* data = reinterpret_cast<OpData*>(node->user_data);
+
+  int64_t num_axis = NumElements(op_context->axis);
+  TfLiteTensor* temp_index;
+  TF_LITE_ENSURE_OK(context,
+                    GetTemporarySafe(context, node, /*index=*/0, &temp_index));
+  TfLiteTensor* resolved_axis;
+  TF_LITE_ENSURE_OK(
+      context, GetTemporarySafe(context, node, /*index=*/1, &resolved_axis));
+  TfLiteTensor* temp_prod;
+  TF_LITE_ENSURE_OK(context,
+                    GetTemporarySafe(context, node, /*index=*/2, &temp_prod));
+
+  const TfLiteTensor* input = op_context->input;
+  TfLiteTensor* output = op_context->output;
+
+  // Return early when input shape has zero dim.
+  for (int i = 0; i < input->dims->size; ++i) {
+    if (input->dims->data[i] == 0) return kTfLiteOk;
+  }
+
+  // Resize the output tensor if the output tensor is dynamic.
+  if (IsDynamicTensor(output)) {
+    TF_LITE_ENSURE_OK(context,
+                      ResizeTempAxis(context, op_context, resolved_axis));
+    TF_LITE_ENSURE_OK(context, ResizeOutputTensor(context, op_context));
+    TF_LITE_ENSURE_OK(context, ResizeTempAccum(context, op_context, temp_prod));
+
+    const int input_size = GetTensorShape(input).FlatSize();
+    const int output_size = GetTensorShape(output).FlatSize();
+    TF_LITE_ENSURE(context, input_size != 0);
+    TF_LITE_ENSURE(context, output_size != 0);
+
+    const int reduced_axis_size = input_size / output_size;
+    const float scaling = GetQuantProdScaling(
+        input->params.scale, output->params.scale, reduced_axis_size);
+    QuantizeMultiplier(scaling, &data->multiplier, &data->shift);
+  }
+
+  TF_LITE_ENSURE(
+      context,
+      reference_ops::QuantizedReduceProd<T>(
+          GetTensorData<T>(input), input->params.zero_point,
+          GetTensorShape(input), GetTensorData<T>(output),
+          output->params.zero_point, GetTensorShape(output),
+          GetTensorData<int>(op_context->axis), num_axis,
+          op_context->params->keep_dims, GetTensorData<int>(temp_index),
+          GetTensorData<int>(resolved_axis), GetTensorData<int32>(temp_prod),
+          data->multiplier, data->shift));
+  return kTfLiteOk;
+}
+
+TfLiteStatus EvalProd(TfLiteContext* context, TfLiteNode* node) {
+  OpContext op_context(context, node);
+  if (op_context.input->quantization.type != kTfLiteNoQuantization) {
```

Comment:
Nit: Can we add a comment to explain the int8 are handled differently by EvalGeneric and EvalQuantizedProd, because int8 can be either a quantized number and a real number?

Reply:
Done